### DISCUSSION TITLE
should-you-use-the-dotnet-8-identity-api-endpoints/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#389] [Announcements]
# should-you-use-the-dotnet-8-identity-api-endpoints/

In this post I discuss the new Identity API endpoints and talk about some of the security and architectural issues they may introduce.

https://andrewlock.net/should-you-use-the-dotnet-8-identity-api-endpoints/

**Comment by JohnGalt1717** [comment]
The biggest fundamental problems:

1. Web browsers need to put in OpenIdConnect native clients that talk to well known endpoints with a client id to get other data about the browser, and then the browser itself secures and encrypts secrets, tokens, refresh tokens etc. This needs to be updated constantly for best practices and forced on devs like Chrome updates.  Javascript should NEVER have access to this data, and should only have a param on fetch for if the token should be passed or not, and the rest including refresh etc. should be done transparently by the browser.  And javascript should never know anything about how authentication works at all other than "here's my well-known endpoint that has all of the info you need in standard format to figure this out for me, now please Log in".

2. The current flow for .net 8 is not secure outside of cookie authentication because it uses effectively a copy of implicit flow in OIDC which has been depreciated because of major security issues.  This should never have seen the light of day. At the very least that should have copied code flow with PKCE.

3. It's just a copy of OIDC without all of the proper tooling of OIDC which they could have added including .well-known, the right endpoints, etc.  It could have been transparent on top of Authentication and Authorization already there in .net without devs having to know ANYTHING about the standard.  They could have spun Blazor pages that can be grabbed with partial code behinds that can be overridden or added on to for UI that don't use any server side and dumped mvc. But no, instead, they got in an ideological war about reinventing the wheel.

So we got another probably 2 versions of .net before they fix the mess they created and do it right.

In the mean time, hopefully someone takes putting in auth providers into browsers seriously and gets it done, because that's how you really fix this mess.

**Comment by paulleblanc** [comment]
The post says there is no way to scaffold the API code into your project, but there is a workaround way to 'dump' the code into your solution. I was unable to find any documentation for how to do this. Can you provide a link that shows how to do this?

**Reply by andrewlock** [reply]
Sure, the workaround is to copy-paste the extensions from GitHub 😅  

https://github.com/dotnet/aspnetcore/blob/2ce54c68b2abfa66974a4e75cf80e61203180ce4/src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs#L57-L63

Obviously you have to keep that all up to date etc and not introduce any further issues, so it's not a great solution

**Reply by JohnGalt1717** [reply]
And what you're really doing is implementing a bad version of OIDC implicit flow which itself is depreciated because of the security issues with it that they haven't addressed.

You're MUCH better off to implement OpenIddict and do it right. 

It's sad that they didn't make OpenIddict a one liner in the framework and pay him for his work, and then update all of the mvc pages to use blazor with auto render mode and create some inheritence patterns that can be overriden to add additional functionality but does best practices on those mvc pages so that styling and layout can be improved without having to break from the embedded pages.

**Reply by kevinchalet** [reply]
@andrewlock great post, as always! :clap:

> And what you're really doing is implementing a bad version of OIDC implicit flow which itself is depreciated because of the security issues with it that they haven't addressed.

Actually, it's not even a bad version of the OIDC implicit flow, it's a clone of the OAuth 2.0 resource owner password credentials grant (aka ROPC) that takes JSON as input instead of formURL encoding (I'm gonna call that fOAuth... for faux OAuth :rofl:)

While discouraged, using ROPC is not fundamentally insecure as long as you only use it with first-party clients (i.e clients you own). Using it with apps maintained by third-party organizations is bad because they have a direct access to the user password, which completely defeats the whole point of using OAuth 2.0.

> It's sad that they didn't make OpenIddict a one liner in the framework and pay him for his work.

Honestly, I would be very happy if they just contributed to the project. Contributing to the documentation (like they did for IdSrv) was mentioned a few times but it never saw the light of day... they just don't care.

I even suggested making these new Identity management API endpoints configurable to be usable with existing projects like IdSrv or OpenIddict (which would benefit everyone), but they didn't seem very interested :rofl:

**Reply by JohnGalt1717** [reply]
I feel your pain Kevin.  This whole thing was adjudicated on Twitter with some seriously bad assumptions by ASP.NET people. I could see the train wreck in process before they wrote a line of code because a position that only huge companies need OIDC and just rerolled stuff that has long since been fixed/improved upon by experts in the field.

**Reply by ProTip** [reply]
And the service extensions are essentially a black-box(without jumping through hoops) to anyone using an IDE that only shows external lib metadata vs fetching and displaying the source(AKA VSCode).

I've observed and come to believe that an over reliance on the service extension pattern can lead to some negative outcomes:

* Reduced quality APIs they build on since they are rarely exercised except by the "experts" creating the extensions or the poor souls responsible for unblocking their teams.
* Under-documented APIs and systems where users have to deduce from example-based documentation and/or read the source code.


I prefer something a little more explicit; composed within each applications code base. Clean and clear on how to modify and compose. Documented excessively somewhere between raw code documentation and example configurations..

**Comment by JohnyL** [comment]
As for point 1, there's `useCookies` paramater you can use to switch to cookie mode. This is excerpt from source code:
```cs
routeGroup.MapPost("/login", async Task<Results<Ok<AccessTokenResponse>, EmptyHttpResult, ProblemHttpResult>>
    ([FromBody] LoginRequest login, [FromQuery] bool? useCookies, [FromQuery] bool? useSessionCookies, [FromServices] IServiceProvider sp)
```
```

**Comment by zejji** [comment]
@JohnGalt1717 @andrewlock @kevinchalet

The principal technical/security criticism (as opposed to architectural) in the article seems to be the use of JSON web tokens stored in the browser. However, can't the risk of this be handled by following the OWASP advice set out here?: https://cheatsheetseries.owasp.org/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.html#token-sidejacking

In summary, the suggestion is to send a random string to the client as a hardened cookie (i.e. secure and HTTP-only) which will be used server-side to verify a SHA256 hash stored in the token. This then prevents the use of the token alone by any malicious actor.

Regarding the other criticisms - from my perspective these seem acceptable trade-offs for many use cases, particularly if it is possible to migrate to a different solution later. I know from bitter experience that the overhead of building in an external auth server and OIDC is immense, and the possibility of a much simpler approach is therefore extremely attractive.

In your opinion, is there any reasonable migration path to, say, using OpenIddict if this is required further down the line? If so, what would this look like at a high level?

Finally, the official ASP.NET Core Identity documentation states that it should be possible to enable users to sign in using OAuth 2.0 with credentials from external authentication providers (e.g. Google, Facebook, Microsoft, Twitter etc.): https://learn.microsoft.com/en-us/aspnet/core/security/authentication/social/?view=aspnetcore-8.0

Just to be clear, are you saying this is incorrect and/or not possible if using the new Identity API endpoints? Again, this is something I will not need initially, but may well need later.

**Reply by kevinchalet** [reply]
> The principal technical/security criticism (as opposed to architectural) in the article seems to be the use of JSON web tokens stored in the browser. However, can't the risk of this be handled by following the OWASP advice set out here?: https://cheatsheetseries.owasp.org/cheatsheets/JSON_Web_Token_for_Java_Cheat_Sheet.html#token-sidejacking

FYI, the new Identity API endpoints don't use JWT tokens but rely on ASP.NET Core Data Protection to produce its tokens (it's the same exact proprietary binary format used by the cookie handler).

> In summary, the suggestion is to send a random string to the client as a hardened cookie (i.e. secure and HTTP-only) which will be used server-side to verify a SHA256 hash stored in the token. This then prevents the use of the token alone by any malicious actor.

Whether the raw token is used or a random hash/reference ID pointing to a DB entry containing the real token doesn't change the threat model at all... because well, the hash becomes the real token 😃

Using HTTP-only cookies to store authentication data helps mitigate token exfiltration as such cookies can't be accessed by JS code. But it's not a perfect protection, despite what the BFF lovers claim: if you have a XSS vulnerability issue, malicious code can still send HTTP requests on behalf of the victim: even if the value can't be accessed by malicious JS code, the authentication cookie will be attached by the browser to every HTTP request sent to the backend, which allows a bad guy to perform any action on behalf of the user.

It's also worth mentioning that many of the recent attacks seen in the wild are executed outside the browser and focus on stealing all the authentication cookies or tokens found in the user data, bypassing any protection available (2FA included).

> I know from bitter experience that the overhead of building in an external auth server and OIDC is immense, and the possibility of a much simpler approach is therefore extremely attractive.

Immense seems a bit excessive to me but if you're looking for a turnkey option, Keycloak makes running your own OIDC server very easy 😃 

> In your opinion, is there any reasonable migration path to, say, using OpenIddict if this is required further down the line? If so, what would this look like at a high level?

Inspired by Andrew's post, I wrote https://kevinchalet.com/2023/10/04/can-you-use-the-asp-net-core-identity-api-endpoints-with-openiddict/, where I demonstrate how you could use use the new Identity API endpoints with OpenIddict as the token generator/validator.

That said, you really shouldn't do that and directly use OpenIddict without these non-configurable API endpoints. Implementing a standard OAuth 2.0 password flow (which is basically what the Identity APIs do) is trivial: https://github.com/openiddict/openiddict-samples/tree/dev/samples/Hollastin/Hollastin.Server

> Just to be clear, are you saying this is incorrect and/or not possible if using the new Identity API endpoints? Again, this is something I will not need initially, but may well need later.

It's not possible at all with the Identity API endpoints, that exclusively support password authentication.

**Reply by JohnGalt1717** [reply]
And I'll add that they did everything except the payloads non-standard.

What this really comes down to is that browser makers MUST implement standards compliant secure authentication systems that take the metadata url, version of the standard and do auth, never show tokens, never share tokens, and only allow the dev to specify sending the token on a given fetch command.  The entire auth/refresh token system including login screen should be opaque to javascript and encrypted and not accessible on the machine directly.

ONLY this, would solve the real problems of security. Everything else is a patch at best.

**Reply by zejji** [reply]
@kevinchalet - Many thanks for your response, much appreciated.

The way I had understood the OWASP suggestion was that *both* the token and the cookie should be required to be sent with any request, i.e. the token contains a SHA256 hash of the random string in the cookie. This allows state to be stored in the token (doing away with the requirement for server-side sessions) and for that state to be accessible to JS, but should prevent simple token sidejacking because neither the token nor the cookie is enough on their own (i.e. the cookie is needed to validate the token).

However, as you say, if both the authentication cookies *and* tokens for a user are stolen, a malicious attacker still has the keys to the kingdom!

My question here is - does implementing a standard OAuth 2.0 password flow (or any other OAuth flow) provide increased protection in this regard? Does it prevent an attack by someone who has both your cookies and your tokens? In what ways does it improve upon the security of the new ASP.NET Core Identity solution?

I'm genuinely interested in getting the security aspect correct, but am finding literature the literature on OAuth/OIDC very difficult to decipher, with many differing opinions on the internet as to what is acceptable!

I probably ought to clarify my particular use case - I am likely the exact target audience for the changes to ASP.NET Core Identity to enable the provision of tokens.

Specifically:
- I have a single front-end Vue 3 SPA talking to a single ASP.NET Core backend server.
- I don't want to use a third-party auth service such as Auth0, Okta, Entra, Cognito or others.
- I currently have no need for either (i) SSO across multiple applications; or (ii) a separate auth server (which is an assumption on which pretty much all of OAuth is premised). SSO may become necessary much further down the line.
- However, I will probably need social login in the medium term.

Ideally I'd like a simple but secure token-based auth solution that supports refresh tokens.

However, it's proving very difficult for me to identify accurately what the best solution is for my use case and whether I would be opening myself up to real security risks by using the new Identity API rather than e.g. OpenIddict with the standard OAuth 2.0 password flow.

**Comment by kevinchalet** [comment]
@JohnGalt1717 if not making the tokens accessible from JS code is important for you, the BFF pattern is a good option. That said, nothing prevents the logged in user himself from sending crafted requests to the BFF itself, that will extract the token from the authentication cookie (or wherever else it's stored) and forward the request to the real API. It's really important to realize that the BFF pattern only prevents token exfiltration and nothing else.

An interesting alternative is to offload the OIDC dance/HTTP requests processing to a service worker that is isolated from the rest of the JS app. It's an approach used by https://github.com/AxaFrance/oidc-client if you're interested in learning more.

> The way I had understood the OWASP suggestion was that both the token and the cookie should be required to be sent with any request, i.e. the token contains a SHA256 hash of the random string in the cookie. This allows state to be stored in the token (doing away with the requirement for server-side sessions) and for that state to be accessible to JS, but should prevent simple token sidejacking because neither the token nor the cookie is enough on their own (i.e. the cookie is needed to validate the token).

@zejji interestingly, what you describe is pretty much how CSRF/antiforgery is implemented, even in ASP.NET Core 😄 

I have personally nothing against cookies, but I believe using antiforgery is a must in this case. Too bad it's not the approach promoted by these new Identity APIs...

> My question here is - does implementing a standard OAuth 2.0 password flow (or any other OAuth flow) provide increased protection in this regard? Does it prevent an attack by someone who has both your cookies and your tokens? In what ways does it improve upon the security of the new ASP.NET Core Identity solution?

OAuth 2.0 doesn't provide any protection against that. The difference with the custom protocol used by the Identity APIs is that there's a specification and a documented threat model that make it very clear.

There was an interesting attempt at mitigating such attacks using TLS token binding, but the Chrome folks decided to remove this feature unilaterally: it couldn't do much to prevent attacks relying on malicious code running on the machine, but it was an efficient and simple way to mitigate token exfiltration (way before BFF was "invented" 🤣)

> However, it's proving very difficult for me to identify accurately what the best solution is for my use case and whether I would be opening myself up to real security risks by using the new Identity API rather than e.g. OpenIddict with the standard OAuth 2.0 password flow.

If you're looking for something simple and don't have cross-domain requirements, using cookies is a good option. But in this case, I'd personally decorate all my API endpoints with `[ValidateAntiforgeryToken]` to enforce antiforgery everywhere. Using authentication cookies without antiforgery is a disaster waiting to happen because it means you're relying on `same-site=lax` or `same-site=strict`: yet, it's possible you'll need at some point to use `same-site=none` (e.g because you want to integrate with an OIDC server that requires `response_mode=form_post`): it's the exact moment you'll shoot yourself in the foot if you don't have another layer like antiforgery to protect your app 😄

**Reply by JohnGalt1717** [reply]
The issue is still that anything other than the browser has access to the tokens. Nothing should. The entire process should be hardened, encrypted and completely opaque other than "Login(metaDataAddress)" and "fetch(withToken)".

And cookies only work if the entire app is hosted on the same URL. It doesn't work as soon as you have a SPA that is on a different domain from the api as a small example.

**Reply by kevinchalet** [reply]
> The issue is still that anything other than the browser has access to the tokens. Nothing should. The entire process should be hardened, encrypted and completely opaque other than "Login(metaDataAddress)" and "fetch(withToken)".

In the real world, that doesn't really work: any process can access the memory of any other process running under the same user session, so even if you encrypt the entire process memory, you still have to store the keys in clear somewhere (any process can access the keys and use them to decrypt the process memory).

As soon as you have malware running on your machine, all hope is lost.

**Reply by JohnGalt1717** [reply]
Sure. To a point.  You have to have Ring 0 access to do so on a modern X64 processor however or direct physical access.

And by that logic ALL authentication is moot.  But what's for sure: Web authentication is broken and it's because browsers are punting their responsibility on the topic.

**Reply by kevinchalet** [reply]
> You have to have Ring 0 access to do so on a modern X64 processor however or direct physical access.

Haha, no. See https://security.stackexchange.com/questions/220771/why-does-windows-allow-non-parent-processes-to-read-write-the-memory-of-a-proces for more information.

FWIW, you can easily dump a process' memory using the Windows Task Manager or Sysinternals' Process Explorer without even requiring administrative rights (unless the target process itself is running under another session, of course).

(also, I only mentioned direct memory access but there are other methods like DLL injection or remote threads).

> And by that logic ALL authentication is moot. But what's for sure: Web authentication is broken and it's because browsers are punting their responsibility on the topic.

It's not an easy thing. When talking about threat models, the most important question is always: what do you want to protect yourself against?
  - If the answer is "malware running on the machine", browsers won't help you with that. Sure you could imagine a browser running in a special enclave isolated by the CPU itself - e.g something-a-la [Intel SGX](https://en.wikipedia.org/wiki/Software_Guard_Extensions) (which is now deprecated in recent consumer products), but I don't see much appetite for that among browser vendors. And such enclaves are not as secure as you might think (tons of vulnerabilities were discovered, which probably explains why Intel decided to deprecate SGX).

  - If the answer is "the user himself", same answer, a user can access his own cookies and do whatever he wants with them, including using them on a different machine.

  - Now, if the answer is just "some malicious JS code injected via XSS", then in most cases, it won't be able to escape the browser's sandbox and we already have tons of options available to reduce the risk. The most important is probably Content Security Policy but surprisingly, it's rarely something developers are familiar with (I'm not even sure it's mentioned in the ASP.NET Core docs 😄)...

**Comment by aryehsilver** [comment]
It looks like a lot of this has changed with the release of .NET 8 I have just created a template, and it is very different to what has been described here.
- It looks like it uses cookies by default and not bearer tokens using `.AddIdentityCookies()`
- The API now only contains 4 endpoints:
  - `PerformExternalLogin`
  - `Logout`
  - `Manage/LinkExternalLogin`
  - `Manage/DownloadPersonalData`

The `Program.cs` file conatins the line: `app.MapAdditionalIdentityEndpoints();` instead of `app.MapGroup("/account").MapIdentityApi<AppUser>();` mentioned here.

All the code to register a new account is done in the `Register.razor` component. When removing the route from that file, it returns a 404 as expected.

It therefore seems that most of these concerns have been dealt with.

Would like to hear your take on this.

**Reply by andrewlock** [reply]
I just tested again with `MapIdentityApi()`, and it includes the same endpoints:

![image](https://github.com/andrewlock/blog-comments/assets/18755388/65644524-7d90-4e46-86e4-1af5f382c945)

`MapAdditionalIdentityEndpoints()` is not part of the Identity APIs. 

I assume you're referring to the `dotnet new blazor -au Individual` template right? That's an entirely different thing to the Identity APIs 🙂 This template uses Blazor Server and a complete reimplementation of the Identity Pages instead of exposing public APIs. The additional links are required for external interactions.

As I understand it, even if you do `dotnet new blazor -au Individual -int WebAssembly` the Identity part is still Blazor Server, so it doesn't expose these endpoints. So basically, the advice of "don't use this" still stands IMO 😄

**Comment by sweeperq** [comment]
Almost every app I have ever written requires user accounts. I'm starting to learn Blazor, so I spent a couple hours trying to understand the new .Net 8 Auth scheme and was almost immediately disappointed. First thing I noticed was that if you are not using the API, the auth razor components are not interactive. They do not validate on the client. If you change the registration or login components to Interactive Server, they breaks everything because HttpContext is inaccessible from razor components using Interactive Server.

So then I heard about the API endpoints and figured I would just put the razor components in the Client project and use InteractiveAuto instead. Then I could have a rich forms experience for the end user. Seemed like a solid plan.

One of the first things I did was add First/Last Name to my AppUser. Quickly found out that you cannot override the "register" endpoint. Then I found out that I couldn't disable or remap the individual endpoints... it is all or nothing.  So if you want to customize your registration API, you basically have to copy the entire contents of the **MapIdentityApi** extension method. Only then can you strip out calls you may not want to support, or customize your registration endpoint to do things like, support usernames that are not email addresses, or add a First/Last Name. But that really isn't even enough. You'll have to re-create any request objects, like **RegisterRequest** to add your custom fields.

Mildly frustrating to have another incomplete Microsoft Identity solution. Everyone keeps telling us we shouldn't roll our own, but we keep getting presented with partially baked solutions. For instance, in addition to the issues already mentioned, where is the user/role/claim admin?

On the plus side, as part of my learning experience I was able to set up Auth0. They provide definitive step-by-step directions on how to set it up. It worked on the very first run. Trying to dig into the specifics of that to see if there is anything I can learn and apply to the Individual Accounts.

**Comment by itfwf** [comment]
Hi Andrew,

Is there a way to combine the endpoints but without a password ? by sending magic links to users email?

**Comment by thirty7four2** [comment]
I've been putting off a move to .net core for years, because of how complicated this topic seems to be. We were using Microsoft.Owin.Security.OAuth, which was not brought over to .net core.

I kept waiting for an official Microsoft solution, but this won't work for us. We only do registration for our site by sending a link to people. Not being able to disable the register api is a deal breaker. I had found openiddict a long time ago, and I kept coming back to it thinking it might be the best solution, but for some reason I still wanted something official from Microsoft.

I've reread so many articles every 6 months or so when I have time to work on upgrading to .net core. Every time it just ended in frustration as there wasn't a clear path for us to take. I considered doing the idea of just taking the code from .net core and customizing it, but it sounded like a big headache. After seeing Kevin's name pop up in so many articles, I finally decided his solution might as well be the "official" solution and decided to use it.

It took a couple of days of trial and error, but I finally got everything working and happy now that we are on .net core 8!

If I have time I will put a sample on github. We are using degraded mode so we could have more control - so we could reuse our tables that we were previously using. The api is just used by a mobile app and Windows app and does the password flow and refresh token.

Thanks @kevinchalet for your work on all this! And, thanks @andrewlock for this great article!

**Reply by kevinchalet** [reply]
Wooo, very nice. Glad you found OpenIddict useful! 😃

(note: OpenIddict is fully compatible with OWIN/Katana, so if you still have APIs that can't be easily migrated to ASP.NET Core, you can modernize them by using the OpenIddict validation handler and its OWIN integration).

Cheers.

**Reply by thirty7four2** [reply]
I added a sample project for this. I mainly just needed a few places to hook in for logging and to use my old tables for access/refresh tokens. I also needed to add a few parameters into the response that my apps depended on. In retrospect, I probably should have just switched to using the provided models to make this easier.

https://github.com/thirty7four2/example_openiddict_password_degradedmode

**Comment by alexanderbikk** [comment]
As usual Andrew you have a brilliant articles with a lot of details I appreciate you work!

However I must admit that only 3-4-5 reasons are indeed an issue especially the 5th. 

I think that the 1st reason is not a issue since you gave the answer - we can enable cookie and use it with SPA apps. For apps with just SPA + API SameSite cookie and disabled CORS should be enough (which is default solution) as I understand. In addition to that we could or probably must configure antiforgery token. Probably the easiest way is to using 2 cookies approach and attach token to every response and then push it back with every request. And the issues with CSRF is solved. Please correct me if I'm wrong.

For the 2nd reason if we decide to add mobile app it's still not an issue we could utilize the same cookie  approach, or enable Bearer token schema. (I'm not really good at mobile apps but it should be more secure to store tokens on devices). And then I can design my Login pages similar to each other. Later I can use any Auth providers, of course, refactoring will be needed but we are not always starting with microservices, k8s, and all other stuff for our applications, so we can grow when needed.

I just wanted to say that in most cases such Auth options would be more than enough for pet projects/simple projects/startups, where using 3rd party Auth server or maintaining you own will cost to much money and/or time. So that instead of being guru of OAuth people can start building value :) We don't always need standalone Auth server and SSO, even if we have more then one app. 

I really don't see any good options for .Net where we have Email confirmations, 2FA and External logins out of the box for affordable price (actually free) without additional hosting etc. Yes, in worst case we will need just copy existing Identity APIs from source and adjust them for our needs since they are not customizable. 

Maybe Supabase or Firebase (I know that 2FA is not supported in .Net SDK) Auth could be an option, however there we still have to implement a lot of things from scratch. 

Finally I would say that this solution is good enough for projects that I mentioned where we need secure enough Auth which we can setup fast. Could it be better at customization certainly it could.

For a bit more complex scenarios I would for sure choose either self-hosted Auth server or external provides.

**Comment by carloswm85** [comment]
Great review. I'm trying to implement Identity API in my own system. I was wondering, how do you think `<PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="10.0.1" />` solves the Bearer Token problem you describe in this post? I'm barely diving into this topic.
